sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala

```diff
@@ -0,0 +1,629 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.{broadcast, TaskContext}
+import org.apache.spark.annotation.{DeveloperApi, Experimental, Unstable}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, SortOrder, SpecializedGetters, UnsafeProjection}
+import org.apache.spark.sql.catalyst.expressions.codegen._
+import org.apache.spark.sql.catalyst.expressions.codegen.Block._
+import org.apache.spark.sql.catalyst.plans.physical.Partitioning
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+import org.apache.spark.sql.execution.vectorized.{OffHeapColumnVector, OnHeapColumnVector, WritableColumnVector}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
+
+/**
+ * :: Experimental ::
+ * Holds a user defined rule that can be used to inject columnar implementations of various
+ * operators in the plan. The [[pre]] function can be used to replace [[SparkPlan]] instances with
+ * versions that support a columnar implementation. After this Spark will insert any transitions
+ * necessary. This includes transitions from row to columnar [[RowToColumnarExec]] and from
+ * columnar to row [[ColumnarToRowExec]].  After this the [[post]] function is called to allow
+ * replacing any of the implementations of the transitions or doing cleanup of the plan, like
+ * inserting stages to build larger batches for more efficient processing, or stages that
+ * transition the data to/from an accelerator's memory.
+ */
+@Experimental
```

Comment:
all apis in execution are non-public so there's no experimental api here.

Reply:
So how do we want to handle this situation where we are exposing some internal implementation details but with no real guarantees of stability?  Not marking them as anything I felt gave the wrong impression that there was some guarantee.

Reply:
They are all private. The package.scala file documents that. Also execution and catalyst packages are ignored in the public API docs.

Basically it's use at your own risk type of internal APIs.